src/java.base/share/classes/java/lang/ThreadLocal.java

```diff
@@ -807,4 +798,43 @@ private void expungeStaleEntries() {
             }
         }
     }
+
+
+    /**
+     * Reads the value of the jdk.traceVirtualThreadLocals property to determine if
+     * a stack trace should be printed when a virtual threads sets a thread local.
+     */
+    private static boolean traceVirtualThreadLocals() {
+        String propValue = GetPropertyAction.privilegedGetProperty("jdk.traceVirtualThreadLocals");
+        return (propValue != null)
+                && (propValue.isEmpty() || Boolean.parseBoolean(propValue));
+    }
+
+    /**
+     * Print a stack trace if the current thread is a virtual thread.
+     */
+    static void dumpStackIfVirtualThread() {
+        if (Thread.currentThread() instanceof VirtualThread vthread) {
+            try {
+                var stack = StackWalkerHolder.STACK_WALKER.walk(s ->
+                        s.skip(1)  // skip caller
+                         .collect(Collectors.toList()));
+
+                // switch to carrier thread to avoid recursive use of thread-locals
```

Comment:
The lambda here uses `System.out` to format and print and there's a loop which converts each collect `StackFrame` to a printable `StackTraceElement`. I had a brief look at the internals of the classes involved, but couldn't spot any code that instantiates or calls `set()` on any `ThreadLocal`s. Is there any specific code here which would trigger recursive thread-local usage or is it more a precaution to prevent any such potential usage.

Reply:
> The lambda here uses `System.out` to format and print and there's a loop which converts each collect `StackFrame` to a printable `StackTraceElement`. I had a brief look at the internals of the classes involved, but couldn't spot any code that instantiates or calls `set()` on any `ThreadLocal`s. Is there any specific code here which would trigger recursive thread-local usage or is it more a precaution to prevent any such potential usage.

Formatting is locale specific and amounts to running arbitrary code, so this cod has to defend against recursive calls.